File: Clifton.WebServer/Router.cs
/*
Copyright (c) 2015, Marc Clifton
All rights reserved.

Article: http://www.codeproject.com/Articles/859108/Writing-a-Web-Server-from-Scratch
Git: https://github.com/cliftonm/BasicWebServer.git

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this list
  of conditions and the following disclaimer. 

* Redistributions in binary form must reproduce the above copyright notice, this 
  list of conditions and the following disclaimer in the documentation and/or other
  materials provided with the distribution. 
 
* Neither the name of MyXaml nor the names of its contributors may be
  used to endorse or promote products derived from this software without specific
  prior written permission. 

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

using Clifton.ExtensionMethods;

namespace Clifton.WebServer
{
	public class ResponsePacket
	{
		public string Redirect { get; set; }
		public byte[] Data { get; set; }
		public string ContentType { get; set; }
		public Encoding Encoding { get; set; }
		public Server.ServerError Error { get; set; }
		public HttpStatusCode StatusCode { get; set; }

		public ResponsePacket()
		{
			Error = Server.ServerError.OK;
			StatusCode = HttpStatusCode.OK;
		}
	}

	public class Route
	{
		public string Verb { get; set; }
		public string Path { get; set; }
		public RouteHandler Handler { get; set; }
		public Func<Session, Dictionary<string, object>, string, string> PostProcess { get; set; }
	}

	internal class ExtensionInfo
	{
		public string ContentType { get; set; }
		public Func<Route, Session, Dictionary<string, object>, string, string, ExtensionInfo, ResponsePacket> Loader { get; set; }
	}

	public class Router
	{
		public string WebsitePath { get; set; }

		public const string POST = "post";
		public const string GET = "get";
		public const string PUT = "put";
		public const string DELETE = "delete";

		private Dictionary<string, ExtensionInfo> extFolderMap;
		private List<Route> routes;

		public Router()
		{
			routes = new List<WebServer.Route>();

			extFolderMap = new Dictionary<string, ExtensionInfo>() 
			{
				{"ico", new ExtensionInfo() {Loader=ImageLoader, ContentType="image/ico"}},
				{"png", new ExtensionInfo() {Loader=ImageLoader, ContentType="image/png"}},
				{"jpg", new ExtensionInfo() {Loader=ImageLoader, ContentType="image/jpg"}},
				{"gif", new ExtensionInfo() {Loader=ImageLoader, ContentType="image/gif"}},
				{"bmp", new ExtensionInfo() {Loader=ImageLoader, ContentType="image/bmp"}},
				{"html", new ExtensionInfo() {Loader=PageLoader, ContentType="text/html"}},
				{"css", new ExtensionInfo() {Loader=FileLoader, ContentType="text/css"}},
				{"js", new ExtensionInfo() {Loader=FileLoader, ContentType="text/javascript"}},
				{"", new ExtensionInfo() {Loader=PageLoader, ContentType="text/html"}},	  // no extension is assumed to be .html
			};
		}

		public void AddRoute(Route route)
		{
			routes.Add(route);
		}

		public ResponsePacket Route(Session session, string verb, string path, Dictionary<string, object> kvParams)
		{
			string ext = path.RightOfRightmostOf('.');
			ExtensionInfo extInfo;
			ResponsePacket ret = null;
			verb = verb.ToLower();
			path = path.ToLower();

			if (extFolderMap.TryGetValue(ext, out extInfo))
			{
				string wpath = path.Substring(1).Replace('/', '\\');			// Strip off leading '/' and reformat as with windows path separator.
				string fullPath = Path.Combine(WebsitePath, wpath);

				Route routeHandler = routes.SingleOrDefault(r => verb == r.Verb.ToLower() && path == r.Path.ToLower());

				if (routeHandler != null)
				{
					// Application has a handler for this route.
					ResponsePacket handlerResponse = null;

					// If a handler exists:
					routeHandler.Handler.IfNotNull((h) => handlerResponse = h.Handle(session, kvParams));

					if (handlerResponse == null)
					{
						// Respond with default content loader.
						ret = extInfo.Loader(routeHandler, session, kvParams, fullPath, ext, extInfo);
					}
					else
					{
						// Respond with redirect.
						ret = handlerResponse;
					}
				}
				else
				{
					// Attempt default behavior
					ret = extInfo.Loader(null, session, kvParams, fullPath, ext, extInfo);
				}
			}
			else
			{
				ret = new ResponsePacket() { Error = Server.ServerError.UnknownType };
			}

			return ret;
		}

		/// <summary>
		/// Read in what is basically a text file and return a ResponsePacket with the text UTF8 encoded.
		/// </summary>
		private ResponsePacket FileLoader(Route routeHandler, Session session, Dictionary<string, object> kvParams, string fullPath, string ext, ExtensionInfo extInfo)
		{
			ResponsePacket ret;

			if (!File.Exists(fullPath))
			{
				ret = new ResponsePacket() { Error = Server.ServerError.FileNotFound };
				Console.WriteLine("!!! File not found: " + fullPath);
			}
			else
			{
				string text = File.ReadAllText(fullPath);
				ret = new ResponsePacket() { Data = Encoding.UTF8.GetBytes(text), ContentType = extInfo.ContentType, Encoding = Encoding.UTF8 };
			}

			return ret;
		}

		/// <summary>
		/// Read in an image file and returns a ResponsePacket with the raw data.
		/// </summary>
		private ResponsePacket ImageLoader(Route routeHandler, Session session, Dictionary<string, object> kvParams, string fullPath, string ext, ExtensionInfo extInfo)
		{
			ResponsePacket ret;

			if (!File.Exists(fullPath))
			{
				ret = new ResponsePacket() { Error = Server.ServerError.FileNotFound };
				Console.WriteLine("!!! File not found: " + fullPath);
			}
			else
			{
				FileStream fStream = new FileStream(fullPath, FileMode.Open, FileAccess.Read);
				BinaryReader br = new BinaryReader(fStream);
				ret = new ResponsePacket() { Data = br.ReadBytes((int)fStream.Length), ContentType = extInfo.ContentType };
				br.Close();
				fStream.Close();
			}

			return ret;
		}

		/// <summary>
		/// Load an HTML file, taking into account missing extensions and a file-less IP/domain, which should default to index.html.
		/// </summary>
		private ResponsePacket PageLoader(Route routeHandler, Session session, Dictionary<string, object> kvParams, string fullPath, string ext, ExtensionInfo extInfo)
		{
			ResponsePacket ret;

			if (fullPath == WebsitePath)		// If nothing follows the domain name or IP, then default to loading index.html.
			{
				ret = Route(session, GET, "/index.html", null);
			}
			else
			{
				if (String.IsNullOrEmpty(ext))
				{
					// No extension, so we make it ".html"
					fullPath = fullPath + ".html";
				}

				// Inject the "Pages" folder into the path
				fullPath = WebsitePath + "\\Pages" + fullPath.RightOf(WebsitePath);

				// Custom, for page not found error.
				if (!File.Exists(fullPath))
				{
					ret = new ResponsePacket() { Error = Server.ServerError.PageNotFound };
					Console.WriteLine("!!! File not found: " + fullPath);
				}
				else
				{
					string text = File.ReadAllText(fullPath);

					// TODO: We put the route custom post process last because of how content is merged in the application's process,
					// but this might cause problems if the route post processor adds something that the app's post processor needs to replace.
					// How do we handle this?  A before/after process?  CSRF tokens are a great example!

					// Do the application global post process replacement.
					text = Server.postProcess(session, fullPath, text);

					// If a custom post process callback exists, call it.
					routeHandler.IfNotNull((r) => r.PostProcess.IfNotNull((p) => text = p(session, kvParams, text)));

					// Do our default post process to catch any final CSRF stuff in the fully merged document.
					text = Server.DefaultPostProcess(session, fullPath, text);


					ret = new ResponsePacket() { Data = Encoding.UTF8.GetBytes(text), ContentType = extInfo.ContentType, Encoding = Encoding.UTF8 };
				}
			}

			return ret;
		}
	}
}

File: Clifton.fsWebServer/Router.fs
module Clifton.fsRouter
// See here: http://fsharpforfunandprofit.com/posts/organizing-functions/
// The above is a shorthand for declaring a namespace and the module name, and the module content does not need to be indented.

// Initial startup code from here: https://sergeytihon.wordpress.com/2013/05/18/three-easy-ways-to-create-simple-web-server-with-f/

// Read about async workflows here: http://en.wikibooks.org/wiki/F_Sharp_Programming/Async_Workflows

open System
open System.Net
open System.Net.Sockets
open System.Text
open System.IO

open Clifton.Extensions

let POST = "post"
let GET = "get"
let PUT = "put"
let DELETE = "delete"

// From: http://stackoverflow.com/questions/13646272/using-functions-before-they-are-declared
// Since functions are first-class objects in F#, you can pass them around instead -- which presents a much nicer (and still immutable) solution than forward references.

// Read about classes, unions, records, and structures: http://msdn.microsoft.com/en-us/library/dd233205.aspx
// We use a record here because we want a reference and we don't need the complexity of a class.

// An enum
type RouterError =
    | OK = 0
    | ExpiredSession = 1
    | NotAuthorized = 2
    | FileNotFound = 3
    | PageNotFound = 4
    | ServerError = 5
    | UnknownType = 6
    | ValidationError = 7

// A record.  Records are annoying because the way the type system infers the
// record type is from the first field and only the first field, therefore,
// we put the the error field first as this is pretty unique.
type ResponsePacket = 
    {
        error : RouterError;
        redirect : string;
        data : byte[];
        contentType : string;
        encoding : Encoding;
    }

// Create a default record so we can clone it and can initialize only the properties we need.
let defaultResponsePacket = {error = RouterError.OK; redirect = null; data = null; contentType = null; encoding = Encoding.UTF8}

// Encode the string into a UTF8 byte array.
let encode (text:string) =
    Encoding.UTF8.GetBytes text

// HTML pages are spoofed to live in a Pages sub-folder rather than directly off the Website folder,
// as this reduces clutter in the Website folder.
let pageLoader (fullPath:string) (ext:string) =
    { defaultResponsePacket with data = File.ReadAllText(fullPath) |> encode; contentType = "text/" + ext}

// Return a ResponsePacket record with the image data.
let imageLoader (fullPath:string) (ext:string) = 
    let fStream = new FileStream(fullPath, FileMode.Open, FileAccess.Read)
    let br = new BinaryReader(fStream)
    let data = br.ReadBytes((int)fStream.Length)
    br.Close()
    fStream.Close()
    {defaultResponsePacket with data = data; contentType = "text/" + ext}

// Return a ResponsePacket with the encoded file data.
let fileLoader (fullPath:string) (ext:string) =
    {defaultResponsePacket with data = File.ReadAllText(fullPath) |> encode; contentType = "text/" + ext}

// If there's no extension, assume html.
// Returns a tuple consisting of the corrected path (including the extension) and the extension itself.
let defaultExtension (path:string) (ext:string) =
    match ext with
    | "" -> (path + ".html", "html")
    | _  -> (path, ext)

let spoofHtmlPath (websitePath:string) ((path:string), (ext:string)) =
    match ext with
    | "html" -> (websitePath + "\\Pages" + path.RightOf(websitePath), ext)
    | _ -> (path, ext)


// Handle the route.  
// TODO: Allow callbacks to the application
// Put the default handler into a function that the application calls if it doesn't handle the route itself.
let route (websitePath:string) (path:string) (verb:string) (kvParams:Map<string, string>) =
    let extAny = path.RightOfRightmostOf('.');
    let wpath = 
        match path with
        | "/"   -> websitePath + @"\index"
        | _     -> websitePath + path.Replace('/', '\\');            // Strip off leading '/' and reformat as with windows path separator.

    let pathAndExt = (defaultExtension wpath extAny) |> (spoofHtmlPath websitePath)
    let fullPath = fst pathAndExt
    let ext = snd pathAndExt

    // Now, if the file exists, we're good to go.
    if File.Exists(fullPath) 
        then 
            match ext with
            | "html" -> pageLoader fullPath ext
            | "ico"  -> imageLoader fullPath ext
            | "png"  -> imageLoader fullPath ext
            | "jpg"  -> imageLoader fullPath ext
            | "gif"  -> imageLoader fullPath ext
            | "bmp"  -> imageLoader fullPath ext
            | "css"  -> fileLoader fullPath ext
            | "js"   -> fileLoader fullPath ext
            | _      -> {defaultResponsePacket with error=RouterError.UnknownType}
        else {defaultResponsePacket with error=RouterError.FileNotFound}
